source/multigrid/mg_transfer_prebuilt.cc

```diff
@@ -326,40 +346,131 @@ void MGTransferPrebuilt<VECTOR>::build_matrices (
+            if (data.size())
+              MPI_Isend(&data[0], data.size()*sizeof(data[0]), MPI_BYTE, dest, 71, tria->get_communicator(), &*requests.rbegin());
+            else
+              MPI_Isend(NULL, 0, MPI_BYTE, dest, 71, tria->get_communicator(), &*requests.rbegin());
+          }
+      }
+
+      // * receive
+      {
+        typename std::vector<dof_pair> receive;
+        for (std::set<unsigned int>::iterator it = neighbors.begin(); it != neighbors.end(); ++it)
+          {
+            MPI_Status status;
+            int len;
+            MPI_Probe(MPI_ANY_SOURCE, 71, tria->get_communicator(), &status);
+            MPI_Get_count(&status, MPI_BYTE, &len);
+
+            if (len==0)
+              {
+                int err = MPI_Recv(NULL, 0, MPI_BYTE, status.MPI_SOURCE, status.MPI_TAG,
+                                   tria->get_communicator(), &status);
+                AssertThrow(err==MPI_SUCCESS, ExcInternalError());
+                continue;
+              }
+
+            int count = len / sizeof(dof_pair);
+            Assert(count * sizeof(dof_pair) == len, ExcInternalError());
+            receive.resize(count);
+
+            void *ptr = &receive[0];
+            int err = MPI_Recv(ptr, len, MPI_BYTE, status.MPI_SOURCE, status.MPI_TAG,
+                               tria->get_communicator(), &status);
+            AssertThrow(err==MPI_SUCCESS, ExcInternalError());
+
+            for (unsigned int i=0; i<receive.size(); ++i)
+              {
+                copy_indices_level_mine[receive[i].level].push_back(
+                  std::make_pair (receive[i].global_dof_index, receive[i].level_dof_index)
+                );
+              }
+          }
+      }
+
+      // * wait
+      if (requests.size() > 0)
+        {
+          MPI_Waitall(requests.size(), &requests[0], MPI_STATUSES_IGNORE);
+          requests.clear();
+        }
+    }
+
   // If we are in debugging mode, we order the copy indices, so we get
   // more reliable output for regression texts
 #ifdef DEBUG
   std::less<std::pair<types::global_dof_index, types::global_dof_index> > compare;
   for (unsigned int level=0; level<copy_indices.size(); ++level)
     std::sort(copy_indices[level].begin(), copy_indices[level].end(), compare);
-  for (unsigned int level=0; level<copy_indices_from_me.size(); ++level)
-    std::sort(copy_indices_from_me[level].begin(), copy_indices_from_me[level].end(), compare);
-  for (unsigned int level=0; level<copy_indices_to_me.size(); ++level)
-    std::sort(copy_indices_to_me[level].begin(), copy_indices_to_me[level].end(), compare);
+  for (unsigned int level=0; level<copy_indices_level_mine.size(); ++level)
+    std::sort(copy_indices_level_mine[level].begin(), copy_indices_level_mine[level].end(), compare);
+  for (unsigned int level=0; level<copy_indices_global_mine.size(); ++level)
+    std::sort(copy_indices_global_mine[level].begin(), copy_indices_global_mine[level].end(), compare);
 #endif
 }
```

Comment:
I believe this function has around 250 lines of code. If you think this is the correct structure, then I'm ok with this -- if you see anything that could usefully be externalized into a separate function, I would not object to that either.